homeassistant/components/google_assistant/helpers.py

```diff
@@ -578,7 +578,6 @@ async def sync_serialize(self, agent_user_id):
             device["customData"] = {
                 "webhookId": self.config.get_local_webhook_id(agent_user_id),
                 "httpPort": self.hass.http.server_port,
-                "httpSSL": self.hass.config.api.use_ssl,
```

Comment:
This value is set to `true` only if a user has configured a certificate with Home Assistant. In that case all communication needs to be encrypted.

Removing this is wrong.

Reply:
So the mDNS record wouldn't point to homeassistant.local?
It would point to the external hostname?
Because unless it points to the external hostname, all it currently does is break local fulfillment...

Reply:
mDNS records point at IP addresses. Google is talking to the device locally. So it needs to know what the settings are of the port directly exposed on Home Assistant. It doesn't care about external address.

 <https://github.com/NabuCasa/home-assistant-google-assistant-local-sdk/blob/master/app.ts#L113-L118>

Reply:
Yes, but unless you happen to have the SSL certificate for the what the mDNS record returns(you wont)
which its `{uuid}.local` and with SSL google would try `https://{uuid}.local:port/` (and I hope its obvious why that will fail)
https://github.com/home-assistant/core/blob/dev/homeassistant/components/zeroconf/__init__.py#L314

I do believe that my solution is no longer correct, but this **will not** work in any way currently, and either zeroconf needs to point to the external hostname, or a second webserver that is local only needs to spin up for things like this.

Reply:
- mDNS always resolves to the IP. Internal and external URL are in the mDNS properties but not used.
- `hass.config.api` is set to the internal IP [here](https://github.com/home-assistant/core/blob/dev/homeassistant/components/http/__init__.py#L194-L203)
- If `hass.config.api.use_ssl` is `True`, it's because the port that Home Assistant is listening on is [created with an SSLContext](https://github.com/home-assistant/core/blob/dev/homeassistant/components/http/__init__.py#L335), meaning only encrypted data is accepted.

Reply:
> * mDNS always resolves to the IP. Internal and external URL are in the mDNS properties but not used.
> * `hass.config.api` is set to the internal IP [here](https://github.com/home-assistant/core/blob/dev/homeassistant/components/http/__init__.py#L194-L203)
> * If `hass.config.api.use_ssl` is `True`, it's because the port that Home Assistant is listening on is [created with an SSLContext](https://github.com/home-assistant/core/blob/dev/homeassistant/components/http/__init__.py#L335), meaning only encrypted data is accepted.

That isn't entirely accurate, but its completely irrelevant anyway, so whether it resolves to the ip or what I said, your still not gonna have a valid ssl certificate for the address mDNS resolves to.

If you don't have a valid cert for the address google attempts to connect over, and `httpSSL` is true, then its going to attempt to connect over HTTPS and when it connects over HTTPS and you don't have a valid cert, it fails.

This isn't theoretical, I had wrote a different service providing local fulfillment, and when the `isSecure` flag is set true in the local fulfillment app(and the `httpSSL` flag directly translates to the `isSecure` flag in the home assistant local fulfilment app), it communicated over https, and because the cert wasn't valid (for the local address) local fulfillment failed.

Reply:
You cannot talk HTTP over a port that only accepts HTTPS. If a certificate is loaded in HA it will only accept encrypted connections. Yes, the cert is not valid but otherwise the data is just ignored. 

Try it out, load a cert in your Home Assistant and then navigate to http://homeassistant.local:8123 instead of https://homeassistant.local

Reply:
I will further unsubscribe from this PR since my comments are deemed "Completely irrelevant".

Reply:
I already said I'm aware my solution isn't correct, but it doesn't change the fact that its still completely broken anyway.

I'll leave this here as documentation when this won't work for users trying to use local fulfillment when they have https enabled. hopefully when you realize I'm right you'll atleast fix the issue.

Reply:
Yes it is broken anyway. But with the current approach if invalid certs were allowed, it could work. With your "solution" it can never work.